### ISSUE TITLE
`.second_to_last` seems to load all records

### ISSUE THREAD

**Issue by aried3r** [rails/rails#32436]
### Steps to reproduce

Using any record you have, `Record.last` loads just the last record from the database, but `Record.second_to_last` seems to load all of them and then using the array operator `[-1]` (guessing).

```Ruby
» Message.last
  Message Load (2.2ms)  SELECT  "messages".* FROM "messages" ORDER BY "messages"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> #<Message:0x00007fba9e01d190> {
                   :id => 15536,
           :created_at => Tue, 27 Mar 2018 14:31:11 CEST +02:00,
           :updated_at => Tue, 27 Mar 2018 14:31:22 CEST +02:00,
}
» Message.second_to_last
  Message Load (45.4ms)  SELECT "messages".* FROM "messages" ORDER BY "messages"."id" ASC
=> #<Message:0x00007fbaa2cc1ca8> {
                   :id => 15535,
           :created_at => Tue, 27 Mar 2018 14:30:11 CEST +02:00,
           :updated_at => Tue, 27 Mar 2018 14:30:16 CEST +02:00,
}
```

Compare to
```Ruby
» Message.last(2).first
  Message Load (0.5ms)  SELECT  "messages".* FROM "messages" ORDER BY "messages"."id" DESC LIMIT $1  [["LIMIT", 2]]
=> #<Message:0x00007fbaa2ce2020> {
                   :id => 15535,
           :created_at => Tue, 27 Mar 2018 14:30:11 CEST +02:00,
           :updated_at => Tue, 27 Mar 2018 14:30:16 CEST +02:00,
}
```

### Expected behavior
`.second_to_last` should not load all records.

### Actual behavior
`.second_to_last` seems to load all records.

### System configuration
**Rails version**: 5.1.6
**Ruby version**: 2.5.1

**Comment by fatkodima** [comment]
Can confirm. Will look at this.

**Comment by rafaelfranca** [comment]
Thanks. This is not an issue, it is more like a feature request so I'm closing. But still feel free to submit a PR.

**Comment by aried3r** [comment]
I slightly disagree, in my opinion having `.first`, `.second`, `.third`, `.last` having completely different performance characteristic than `.second_to_last`, `.third_to_last` is definitely an issue and unexpected behaviour.

While the [docs](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-second_to_last) seem to correctly suggest everything is loaded:
```Ruby
Person.second_to_last # returns the second-to-last object fetched by SELECT * FROM people
```
The same is said for [`.second`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-second):
```Ruby
Person.second # returns the second object fetched by SELECT * FROM people
```
but indeed it is using `LIMIT` and `OFFSET`
```Ruby
Person.second # SELECT  "people".* FROM "people" ORDER BY "people"."id" ASC LIMIT 1 OFFSET 1
```
much like the documentation for [`.first`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-first) correctly implies:
```
Person.first # returns the first object fetched by SELECT * FROM people ORDER BY people.id LIMIT 1
```

So I think the docs also need some work, but this is definitely a performance issue and not obvious in usage, since I don't think any linters pick this up yet.

**Comment by fatkodima** [comment]
This is working differently on master (after merging this commit https://github.com/rails/rails/commit/f5328d9379da473bfc5420996693a14c6e0d2556). Probably,  that should be backported to 5.1.x?

**Comment by rafaelfranca** [comment]
That is a performance improvement so usually we don't backport to stable releases. That commit is included in 5.2 though

**Comment by aried3r** [comment]
Hi, I didn't want to open another issue for this, but I saw that in Rails 5.2 this is indeed behaving as expected:

```Ruby
Message.last
  Message Load (0.6ms)  SELECT  "messages".* FROM "messages" ORDER BY "messages"."id" DESC LIMIT $1  [["LIMIT", 1]]

Message.second_to_last
  Message Load (0.5ms)  SELECT  "messages".* FROM "messages" ORDER BY "messages"."id" DESC LIMIT $1  [["LIMIT", 2]]
```

However this is only with an implicit order, I still get the same poorly performing query using an explicit oder:
```Ruby
Message.order(:created_at).last
  Message Load (4.5ms)  SELECT  "messages".* FROM "messages" ORDER BY "messages"."created_at" DESC LIMIT $1  [["LIMIT", 1]]

Message.order(:created_at).second_to_last
  Message Load (71.3ms)  SELECT "messages".* FROM "messages" ORDER BY "messages"."created_at" ASC
```

Since you said this is not an issue but a feature request, what would be the best place to write it down somewhere? I'd also love to improve the performance myself and submit a PR, but I'm not sure where to start or to look, do you have any pointers?
